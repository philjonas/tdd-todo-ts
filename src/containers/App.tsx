import { observer } from "mobx-react";
import React  from "react";
import { TaskAdder } from "../components/TaskAdder";
import { TaskItem } from "../components/TaskItem";
import {
  addItem,
  completeItemToggle,
  deleteItem,
  StoreProps,
} from "../state/Store";
import "./App.css";

export const App = ({ store }: { store: StoreProps }): JSX.Element => {
  const onDelete = (id: number) => deleteItem(store, id);
  const onComplete = (id: number) => completeItemToggle(store, id);
  const handleAdd = (value: string) => addItem(store, value);

  const TodosView = observer(() => (
    <div data-testid="app-todo-container">
      {store.data.map((item) => (
        <TaskItem
          key={item.id}
          {...item}
          onComplete={onComplete}
          onDelete={onDelete}
        />
      ))}
    </div>
  ));
  const TaskView = observer(() => <TaskAdder addTask={handleAdd} />);

  return (
    <div className="App">
      <h1>Ye Olde ToDo App</h1>
      <TaskView />
      <TodosView />
    </div>
  );
};
