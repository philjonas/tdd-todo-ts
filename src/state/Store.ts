import { observable, action } from "mobx";
import { TaskItemProps } from "../components/TaskItem";

export interface StoreProps {
  itemId: number;
  data: TaskItemProps[];
}

export const initialStore: StoreProps = {
  itemId: 3,
  data: [
    { id: 1, value: "Learn TDD", isCompleted: false },
    { id: 2, value: "Write TDD", isCompleted: false },
    { id: 3, value: "Peace of mind", isCompleted: false },
  ],
};

export const store: StoreProps = observable({ ...initialStore });

export const addItem = action((state: StoreProps, value: string): void => {
  if (!value) {
    return;
  }
  const newId = state.itemId + 1;
  state.itemId = newId;
  state.data = [
    {
      id: newId,
      value,
      isCompleted: false,
    },
    ...state.data,
  ];
});

export const completeItemToggle = action(
  (state: StoreProps, id: number): void => {
    const dataClone = state.data.map((item) => {
      if (item.id === id) {
        return { ...item, isCompleted: !item.isCompleted };
      }
      return item;
    });
    state.data = [...dataClone];
  }
);

export const deleteItem = action((state: StoreProps, id: number): void => {
  const dataClone = state.data.filter((item) => item.id !== id);
  state.data = [...dataClone];
});
