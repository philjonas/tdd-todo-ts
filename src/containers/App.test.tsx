import { render, fireEvent, getByTestId } from "@testing-library/react";
import { observable } from "mobx";
import { initialStore, store, StoreProps } from "../state/Store";
import { App } from "./App";

const setup = () => {
  const setupStore: StoreProps = observable({ ...initialStore });
  const utils = render(<App store={setupStore} />);
  const input = utils.getByTestId("task-adder-text");
  const add = utils.getByTestId("task-adder-add");
  const list = utils.getByTestId("app-todo-container");
  return {
    input,
    add,
    list,
    ...utils,
  };
};

describe("App", () => {
  it("renders correctly", () => {
    const wrapper = <App store={store} />;
    expect(wrapper).toMatchSnapshot();
  });

  it("adds a new task", () => {
    const { list, input, add } = setup();
    expect(list.children.length).toBe(3);
    fireEvent.change(input, { target: { value: "123" } });
    expect(input.getAttribute("value")).toBe("123");
    fireEvent.click(add);
    expect(list.children.length).toBe(4);
  });

  it("does not add a new task when the input is blank", () => {
    const { list, input, add } = setup();
    expect(list.children.length).toBe(3);
    fireEvent.change(input, { target: { value: "" } });
    expect(input.getAttribute("value")).toBe("");
    fireEvent.click(add);
    expect(list.children.length).toBe(3);
  });

  it("gets completed on click", () => {
    const { list } = setup();
    expect(list.children.length).toBe(3);
    const item = list.children.item(0);
    const completeBTN = getByTestId(item as HTMLElement, "task-item-completed");
    const txt = getByTestId(item as HTMLElement, "task-item-text");
    expect(txt.className).toBe("task-item-text");
    fireEvent.click(completeBTN);
    expect(txt.className).toBe("task-item-text completed");
  });

  it("gets deleted on click", () => {
    const { list } = setup();
    expect(list.children.length).toBe(3);
    const item = list.children.item(0);
    const deleteBTN = getByTestId(item as HTMLElement, "task-item-delete");
    fireEvent.click(deleteBTN);
    expect(list.children.length).toBe(2);
  });
});
