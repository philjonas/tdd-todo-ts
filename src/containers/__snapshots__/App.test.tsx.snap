// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders correctly 1`] = `
<App
  store={
    Object {
      "data": Array [
        Object {
          "id": 1,
          "isCompleted": false,
          "value": "Learn TDD",
        },
        Object {
          "id": 2,
          "isCompleted": false,
          "value": "Write TDD",
        },
        Object {
          "id": 3,
          "isCompleted": false,
          "value": "Peace of mind",
        },
      ],
      "itemId": 3,
    }
  }
/>
`;
